File: lib/models/care_routine_template_entity.dart
class CareRoutineItem {
  final String type;
  final String time;
  final String title;
  final String? description;

  CareRoutineItem({
    required this.type,
    required this.time,
    required this.title,
    this.description,
  });

  factory CareRoutineItem.fromMap(Map<String, dynamic> m) => CareRoutineItem(
    type: (m['type'] ?? '').toString(),
    time: (m['time'] ?? '').toString(),
    title: (m['title'] ?? '').toString(),
    description: (m['description'] ?? '').toString().isEmpty ? null : m['description'].toString(),
  );

  Map<String, dynamic> toMap() => {
    'type': type,
    'time': time,
    'title': title,
    if (description != null && description!.isNotEmpty) 'description': description,
  };
}

class CareRoutineTemplateEntity {
  final String? id;
  final String name;
  final String description;
  final List<CareRoutineItem> items;
  final String createdBy;
  final String createdAtIso;

  CareRoutineTemplateEntity({
    this.id,
    required this.name,
    required this.description,
    required this.items,
    required this.createdBy,
    required this.createdAtIso,
  });

  factory CareRoutineTemplateEntity.fromFirestore(String id, Map<String, dynamic> data) {
    final list = (data['items'] as List? ?? const []);
    return CareRoutineTemplateEntity(
      id: id,
      name: (data['name'] ?? '').toString(),
      description: (data['description'] ?? '').toString(),
      items: list.map((e) => CareRoutineItem.fromMap(Map<String, dynamic>.from(e as Map))).toList(),
      createdBy: (data['createdBy'] ?? '').toString(),
      createdAtIso: (data['createdAt'] ?? data['createdAtIso'] ?? '').toString(),
    );
  }

  Map<String, dynamic> toFirestore() => {
    'name': name,
    'description': description,
    'items': items.map((i) => i.toMap()).toList(),
    'createdBy': createdBy,
    'createdAt': createdAtIso,
    'lastUpdatedAt': DateTime.now().toIso8601String(),
  };

  List<String> validate() {
    final errs = <String>[];
    if (name.trim().isEmpty) errs.add('Template name is required');
    if (items.isEmpty) errs.add('Please add at least one activity to the template');
    for (int i = 0; i < items.length; i++) {
      final it = items[i];
      if (it.time.trim().isEmpty) errs.add('Activity ${i + 1}: Time is required');
      if (it.title.trim().isEmpty) errs.add('Activity ${i + 1}: Title is required');
    }
    return errs;
  }
}

File: src/entity/careRoutineTemplateEntity.js
export class careRoutineTemplateEntity {
  constructor({
    id = null,
    name = '',
    description = '',
    items = [],
    createdBy = '',
    createdAt = new Date().toISOString(),
    lastUpdatedAt = new Date().toISOString(),
    isActive = true
  } = {}) {
    this.id = id;
    this.name = name;
    this.description = description;
    this.items = items;
    this.createdBy = createdBy;
    this.createdAt = createdAt;
    this.lastUpdatedAt = lastUpdatedAt;
    this.isActive = isActive;
  }

  validate() {
    const errors = [];
    
    if (!this.name || this.name.trim().length === 0) {
      errors.push('Template name is required');
    }
    
    if (this.items.length === 0) {
      errors.push('At least one routine item is required');
    }
    
    for (let i = 0; i < this.items.length; i++) {
      const item = this.items[i];
      if (!item.time || !item.title) {
        errors.push(`Item ${i + 1} requires both time and title`);
      }
      
      if (!this.isValidTime(item.time)) {
        errors.push(`Item ${i + 1} has an invalid time format`);
      }
    }
    
    return errors;
  }

  isValidTime(time) {
    const timeRegex = /^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/;
    return timeRegex.test(time);
  }

  toFirebase() {
    return {
      name: this.name,
      description: this.description,
      items: this.items,
      createdBy: this.createdBy,
      createdAt: this.createdAt,
      lastUpdatedAt: this.lastUpdatedAt,
      isActive: this.isActive
    };
  }

  static fromFirebase(id, data) {
    return new careRoutineTemplateEntity({
      id,
      name: data.name,
      description: data.description,
      items: data.items || [],
      createdBy: data.createdBy,
      createdAt: data.createdAt,
      lastUpdatedAt: data.lastUpdatedAt,
      isActive: data.isActive !== undefined ? data.isActive : true
    });
  }
}